ibis/expr/datatypes/core.py

```diff
@@ -808,12 +826,16 @@ def _pretty_piece(self) -> str:
         return f"<{self.value_type}>"
 
 
+K = TypeVar("K", bound=DataType, covariant=True)
+V = TypeVar("V", bound=DataType, covariant=True)
```

Comment:
Should these be covariant?

As you have it here `Map[Int8, Int32]` is a subtype of `Map[Integer, Integer]`, and I'm not sure we should make that assumption just yet.

Reply:
Even if we consider these as immutable? Actually the validation system will raise unless we define the type vars as covariant.

Reply:
Let's leave it as you have it for now.